### DISCUSSION TITLE
在播放列表，使用倍速快捷键，最高倍速只能够2倍了。

### DISCUSSION THREAD

**Discussion by kethoven** [the1812/Bilibili-Evolved#1357] [综合讨论 / General]
脚本最新版本，在稍后再看播放页面，使用倍速快捷键，最高倍速只能够2倍了。
而且，在不记忆播放速度情况下，播放列表播放下一个视频时候，会显示上一个播放视频的倍速。与实际倍速不相符。

**Comment by LonelySteve** [comment]
> 在稍后再看播放页面，使用倍速快捷键，最高倍速只能够2倍了。

这确实是 BUG，合集播放页（收藏夹，稍后再看）都有这样的问题

> 在不记忆播放速度情况下，播放列表播放下一个视频时候，会显示上一个播放视频的倍速。与实际倍速不相符。

原版就是这样设定的，你可以关掉脚本看看

**Comment by LonelySteve** [comment]
@the1812 

有什么好的办法能在稍后再看，收藏夹等合集播放页观察到分 p 的变化吗？

归根结底，`Observer.videoChange` 没有被正确地实现，因为像这些合集播放页都没有确定的 `cid`，最终导致记忆播放速度功能在这些页面失效（

**Reply by the1812** [reply]
我只知道对于普通视频, `cid`是目前最靠谱的方法 (以前各种监听DOM元素, 巨坑)
合集页我之后看下有没有什么比较好的监听目标

**Reply by LonelySteve** [reply]
我发现这个合集页其实是会改 `cid` 的，但首次加载页面的时候值为 `null`，或许可以判断一下 url 特征，针对这些特别的页面绕过 `Observer.videoChange` 来完成功能的初始化工作。

**Reply by the1812** [reply]
只要要让`Observer.videoChange`能够拿到初始`cid`就好了, 这个播单页面的代码我们可以看下, 正好他留了 SourceMap

**Reply by LonelySteve** [reply]
研究了亿下，可能最好的方式是在 Observer.videoChange 里轮询 `document.URL` 的参数 `bvid` 的变化

合集页加载视频，相应的 bvid 是会作为路由参数写在  `document.URL`  里的（哪怕是首次加载没有传入，也会回写）：

![image](https://user-images.githubusercontent.com/34429322/103000849-cc35f100-4566-11eb-9040-3a7feb8350f6.png)

想拿到初始的 `cid`，估计只能再发一次请求

![image](https://user-images.githubusercontent.com/34429322/103001027-2767e380-4567-11eb-94bc-ed3ecb323bdd.png)

```javascript
 static async videoChange(callback) {
    const ids = await Promise.race([
      SpinQuery.select(() => unsafeWindow.cid).then(v => ({ cid: v })),
      SpinQuery.select(() => {
        const result = document.URL.match(/https:\/\/www\.bilibili\.com\/medialist\/play\/.*(BV\w+)/)
        return result ? result[1] : null
      }).then(v => ({ bvid: v }))
    ])

    if (ids.cid === null && ids.bvid === null) {
      return
    }
    if (!cidHooked) {
      let hookedCid = ids.cid
      Object.defineProperty(unsafeWindow, 'cid', {
        get () {
          return hookedCid
        },
        set (newId) {
          hookedCid = newId
          if (!Array.isArray(newId)) {
            videoChangeCallbacks.forEach(it => it())
          }
        }
      })
      cidHooked = true
    }
    if (ids.bvid) {
      callback();
    }
    const videoContainer = await SpinQuery.select('#bilibili-player video')
    if (videoContainer) {
      Observer.childList(videoContainer, callback)
    } else {
      callback()
    }
    videoChangeCallbacks.push(callback)
  }
```

像上面这样写，应该可以解决此贴的 BUG，但涉及到 `rememberVideoSpeed` 这块的功能还是会有问题

**Reply by the1812** [reply]
我找到一个不用发请求的
```js
unsafeWindow.player.getVideoMessage()
```

**Reply by LonelySteve** [reply]
好活！那 `Observer.videoChange` 就用这个接口来实现吧，而且看起来普通视频页也能用这个接口

**Reply by the1812** [reply]
那我先直接改了, 正好一会有个新版本要发

**Reply by LonelySteve** [reply]
我稍微测试了一下，不知道是不是因为观察时机不对，倍数更新总是延迟了两三秒，而且 `rememberVideoSpeed` 因为用到了 `unsafeWindow.aid` 因此在首次加载时是有问题的

**Reply by the1812** [reply]
aid 当然也会设置的
```JavaScript
static async videoChange (callback) {
  const cid = await SpinQuery.select(() => {
    if (unsafeWindow.cid) {
      return unsafeWindow.cid
    }
    if (unsafeWindow.player?.getVideoMessage) {
      const info = unsafeWindow.player.getVideoMessage()
      if (Number.isNaN(info.cid)) {
        return null
      }
      if (!unsafeWindow.aid) {
        unsafeWindow.aid = info?.aid?.toString()
      }
      return info?.cid?.toString()
    }
  })
  if (cid === null) {
    return
  }
  if (!cidHooked) {
    let hookedCid = cid
    Object.defineProperty(unsafeWindow, 'cid', {
      get () {
        return hookedCid
      },
      set (newId) {
        hookedCid = newId
        if (!Array.isArray(newId)) {
          videoChangeCallbacks.forEach(it => it())
        }
      }
    })
    cidHooked = true
  }
  // callback();
  const videoContainer = await SpinQuery.select('#bilibili-player video')
  if (videoContainer) {
    Observer.childList(videoContainer, callback)
  } else {
    callback()
  }
  videoChangeCallbacks.push(callback)
}
```

**Reply by LonelySteve** [reply]
把

`const videoContainer = await SpinQuery.select('#bilibili-player video')`

改成：

`const videoContainer = await SpinQuery.select('#bilibili-player video, .bilibili-player-video video')`

**Reply by the1812** [reply]
`.bilibili-player-video video`这个普通视频是不是也有, 有的话我直接替换掉`#bilibili-player video`

**Reply by LonelySteve** [reply]
是的，直接用 `.bilibili-player-video video` 就可以了
